src/bio.c

```diff
@@ -278,6 +293,158 @@ void *bioProcessBackgroundJobs(void *arg) {
+                     * convert "\r\n" to '\0' so 1 byte is lost. */
+                    server.bio_stat_net_repl_input_bytes += nread + 1;
+                }
+                int ret = inspectBulkPayloadHeaderForErrors(buf);
+                if (ret == INSPECT_BULK_PAYLOAD_PRIMARY_ABORT) {
+                    replicaBioSaveServerLog(LL_WARNING, "PRIMARY aborted replication with an error: %s", buf + 1);
+                    error = 1;
+                    goto done;
+                } else if (ret == INSPECT_BULK_PAYLOAD_PRIMARY_PING) {
+                    atomic_store_explicit(&server.repl_transfer_lastio, atomic_load_explicit(&server.unixtime, memory_order_relaxed), memory_order_relaxed);
+                    memset(buf, 0, PROTO_IOBUF_LEN);
+                    received_ping = 1;
+                } else if (ret == INSPECT_BULK_PAYLOAD_PRIMARY_BAD_PROTO) {
+                    replicaBioSaveServerLog(LL_WARNING,
+                                            "Bad protocol from PRIMARY, the first byte is not '$' (we received '%s'), are you sure the host "
+                                            "and port are right?",
+                                            buf);
+                    error = 1;
+                    goto done;
+                }
+            } while (received_ping);
+
+            usemark = inspectBulkPayloadHeaderForEOF(buf, eofmark, lastbytes);
+            if (usemark) {
+                server.bio_repl_transfer_size = 0;
+                replicaBioSaveServerLog(LL_NOTICE, "PRIMARY <-> REPLICA sync: receiving streamed RDB from primary with EOF to disk");
+            } else {
+                server.bio_repl_transfer_size = strtol(buf + 1, NULL, 10);
+                replicaBioSaveServerLog(LL_NOTICE, "PRIMARY <-> REPLICA sync: receiving %lld bytes from primary to disk",
+                                        (long long)server.bio_repl_transfer_size);
+            }
+
+            while (1) {
+                if (shouldAbortSave()) {
+                    main_thread_abort = 1;
+                    goto done;
+                }
+                if (usemark) {
+                    readlen = sizeof(buf);
+                } else {
+                    left = server.bio_repl_transfer_size - server.bio_repl_transfer_read;
+                    readlen = (left < (signed)sizeof(buf)) ? left : (signed)sizeof(buf);
+                }
+                nread = connRead(conn, buf, readlen);
+                if (nread <= 0) {
+                    if (connGetState(conn) == CONN_STATE_CONNECTED) {
+                        /* equivalent to EAGAIN */
+                        memset(buf, 0, PROTO_IOBUF_LEN);
+                        continue;
+                    }
+                    replicaBioSaveServerLog(LL_WARNING, "I/O error trying to sync with PRIMARY: %s",
+                                            (nread == -1) ? connGetLastError(conn) : "connection lost");
+                    error = 1;
+                    goto done;
```

Comment:
why not to create an `error` label instead of using a variable?

Reply:
I think we cannot have more than one label inside the `else if (job_type == BIO_RDB_SAVE)`.
If we have 
```
done:
...
error:
...
```